[jit] add simple inheritance support to AST

suo opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #23098 [wip][jit] serialize modules as classes
* #23154 [jit] Put all modules in the global Python CU
* #23170 [jit] Make optimize a thread_local flag
* **#23109 [jit] add simple inheritance support to AST**
* #23097 [jit] kill module_lookup
* #23033 [jit] make RHS  of assignment optional
* #23031 [jit] Parse all stmts in class defs

As title. This is the bare minimum necessary to get inheritance from
torch.nn.Module working. It is guarded everywhere because we don't
support inheritance in a user-visible way.

gh-metadata: pytorch pytorch 23109 gh/suo/100/head

Differential Revision: [D16441914](https://our.internmc.facebook.com/intern/diff/D16441914)

suo committed:
[jit] add simple inheritance support to AST

As title. This is the bare minimum necessary to get inheritance from
torch.nn.Module working. It is guarded everywhere because we don't
support inheritance in a user-visible way.

zdevito commented on torch/csrc/jit/import_source.cpp:
```diff
@@ -190,6 +190,9 @@ struct SourceImporter {
       auto parsed_treeref = p_.parseClassLike();
       if (parsed_treeref->kind() == TK_CLASS_DEF) {
         auto class_def = ClassDef(parsed_treeref);
+        TORCH_CHECK(
```
ErrorReport? class_def has a location at which this error should be reported.

zdevito commented on torch/csrc/jit/script/init.cpp:
```diff
@@ -736,6 +736,9 @@ void initJitScriptBindings(PyObject* module) {
          const ClassDef& classDef,
          ResolutionCallback rcb) {
         C10_LOG_API_USAGE_ONCE("torch.script.class");
+        TORCH_CHECK(
```
ErrorReport, similar to above.

zdevito commented on torch/csrc/jit/script/parser.cpp:
```diff
@@ -649,19 +649,21 @@ struct ParserImpl {
         return parseNamedTuple(name);
       } else {
-        L.reportError("Inheritance is not supported for TorchScript classes");
+        superclass = Maybe<Ident>::create(id.range, Ident::create(id.range, id.text()));
```
We should probably serialize with fully qualified names for modules, meaning this will be torch.nn.Module not `Module` (and hence not an identifier). Why? What happens when a user manages to call their own module Module. Since we do not allow inheritance now, it can't happen, but I can imagine it happening in the future.

suo committed:
Update on "[jit] add simple inheritance support to AST"

[jit] add simple inheritance support to AST

As title. This is the bare minimum necessary to get inheritance from
torch.nn.Module working. It is guarded everywhere because we don't
support inheritance in a user-visible way.

gh-metadata: pytorch pytorch 23109 gh/suo/100/head

suo committed:
Update on "[jit] add simple inheritance support to AST"

[jit] add simple inheritance support to AST

As title. This is the bare minimum necessary to get inheritance from
torch.nn.Module working. It is guarded everywhere because we don't
support inheritance in a user-visible way.

gh-metadata: pytorch pytorch 23109 gh/suo/100/head

suo committed:
Update on "[jit] add simple inheritance support to AST"

[jit] add simple inheritance support to AST

As title. This is the bare minimum necessary to get inheritance from
torch.nn.Module working. It is guarded everywhere because we don't
support inheritance in a user-visible way.

gh-metadata: pytorch pytorch 23109 gh/suo/100/head
